<a href="https://colab.research.google.com/github/lasyaEd/MovieReviewAnalysis/blob/main/02_Methodology.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Sentiment Analysis On IMDB Movie Review - "Gone with the wind"

In [163]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [164]:
file_path = '/content/drive/My Drive/goneWithTheWind.xlsx'

In [165]:
import pandas as pd
df = pd.read_excel(file_path)
df.head()

Review_Date                Author  Rating  \
0  20 November 2017  paskuniag-584-890551     9.0   
1  10 November 2019           daisukereds     8.0   
2    7 January 2005      angel_de_tourvel     8.0   
3  15 December 2019          IshtiaqAhmed    10.0   
4      12 June 2020               Turanic    10.0   

                                        Review_Title  \
0                    Like the film? Read the book.\n   
1     Even after years of being made.. it is good!\n   
2             An immortal and towering achievement\n   
3   Miracle in Film Making - Can't believe they d...   
4      The tragedy of Scarlet the wise 2020 review\n   

                                              Review  \
0  I've seen the film many times, have always enj...   
1  I wish people made more movies like this nowad...   
2  It is always in people's nature to put down gr...   
3  This movie was on my watch list since 1996 or ...   
4  At first you would think, the main theme of th...   

                       Review_Url  
0  /review/rw9937224/?ref_=tt_urv  
1  /review/rw9937224/?ref_=tt_urv  
2  /review/rw9937224/?ref_=tt_urv  
3  /review/rw9937224/?ref_=tt_urv  
4  /review/rw9937224/?ref_=tt_urv

In [170]:
# Drop rows where 'Rating' is NaN
imdb_data = df.dropna(subset=['Rating'])
imdb_data = df.drop(columns=['Review_Url'])
# Create the 'y' column with sentiment based on the rating
imdb_data['y'] = imdb_data['Rating'].apply(lambda x: 'positive' if x > 5 else 'negative')

imdb_data.head(20)

Review_Date                Author  Rating  \
0    20 November 2017  paskuniag-584-890551     9.0   
1    10 November 2019           daisukereds     8.0   
2      7 January 2005      angel_de_tourvel     8.0   
3    15 December 2019          IshtiaqAhmed    10.0   
4        12 June 2020               Turanic    10.0   
5    22 February 2005              jotix100    10.0   
6       5 August 2006         planktonrules    10.0   
7     21 October 2006            bkoganbing    10.0   
8         8 July 2011           murtaza_mma    10.0   
9        26 June 2004              davergod     NaN   
10    18 October 2023                vryaak     6.0   
11   15 November 2003            Alexis_Ray    10.0   
12       7 March 2016             trekhyon5     7.0   
13    8 February 2000            grim_angel     3.0   
14  27 September 2006                dhable    10.0   
15   13 December 2000              mikazuki    10.0   
16      30 March 2022               PCC0921     9.0   
17      26 April 2008    Nazi_Fighter_David     9.0   
18      28 April 2022       jamesjustice-92    10.0   
19       11 June 2020       moorekarl-43260    10.0   

                                         Review_Title  \
0                     Like the film? Read the book.\n   
1      Even after years of being made.. it is good!\n   
2              An immortal and towering achievement\n   
3    Miracle in Film Making - Can't believe they d...   
4       The tragedy of Scarlet the wise 2020 review\n   
5                                Georgia on my mind\n   
6       About as good as it gets in 1930s Hollywood\n   
7         Scarlett's So High Spirited And Vivacious\n   
8    A Potpourri of Vestiges Review: An American m...   
9    A few flaws, but undeniably a massive achieve...   
10     A Timeless Movie About Love and Being Strong\n   
11   Astounding Film - GWTW is the Definition of a...   
12   A classic of respectable scope, well ahead of...   
13   Welcome to the most overrated film in history.\n   
14      The Greatest Film of its Time, and All Time\n   
15        A Classic in the History of Movie-making.\n   
16   Margaret Mitchell won the Pulitzer Prize for ...   
17                          A rich romantic film...\n   
18      There's never going to be a movie like this\n   
19                                   Amazing movie.\n   

                                               Review         y  
0   I've seen the film many times, have always enj...  positive  
1   I wish people made more movies like this nowad...  positive  
2   It is always in people's nature to put down gr...  positive  
3   This movie was on my watch list since 1996 or ...  positive  
4   At first you would think, the main theme of th...  positive  
5   This film shows the best of the American cinem...  positive  
6   I've done so many reviews on IMDb and despite ...  positive  
7   Before I ever saw Gone With the Wind, I was we...  positive  
8   Margaret Mitchell's Gone with the Wind, in its...  positive  
9   One of the greatest achievements in film histo...  negative  
10  "Gone with the Wind" is a well-known film that...  positive  
11  The setting is a Georgia plantation. The year ...  positive  
12  So, I've been calling myself a movie buff for ...  positive  
13  I have yet to figure out the allure of this fi...  negative  
14  I believe that when one views a film, one shou...  positive  
15  Every time I watch this film, and I've seen it...  positive  
16  The year 1939 was a busy year for the Hollywoo...  positive  
17  Gerard O'Hara (Thomas Mitchell), an Irish immi...  positive  
18  It was way back on New Year's day of 2014 when...  positive  
19  An absolute classic of a film that everyone lo...  positive

In [171]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import nltk
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelBinarizer
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
from wordcloud import WordCloud,STOPWORDS
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize,sent_tokenize
from bs4 import BeautifulSoup
import spacy
import re,string,unicodedata
from nltk.tokenize.toktok import ToktokTokenizer
from nltk.stem import LancasterStemmer,WordNetLemmatizer
from sklearn.linear_model import LogisticRegression,SGDClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from textblob import TextBlob
from textblob import Word
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
import os
import warnings

In [172]:
imdb_data.describe()

Rating
count  863.000000
mean     8.355736
std      2.459366
min      1.000000
25%      8.000000
50%     10.000000
75%     10.000000
max     10.000000

In [173]:
imdb_data['y'].value_counts()

y
positive    757
negative    262
Name: count, dtype: int64

### Text Preprocessing

#### Text normalization:
*Tokenization involves splitting a statement into individual words using the word_tokenize() method.*


In [174]:
import nltk
nltk.download('stopwords') # Download the stopwords corpus

tokenizer=ToktokTokenizer()
#Setting English stopwords
stopword_list=nltk.corpus.stopwords.words('english')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### Removing special characters

*To handle English-language evaluations, we must remove any special characters from the dataset.*

In [175]:
#Define function for removing special characters
def remove_special_characters(text, remove_digits=True):
    pattern=r'[^a-zA-z0-9\s]'
    text=re.sub(pattern,'',text)
    return text
#Apply function on review column
imdb_data['Review']=imdb_data['Review'].apply(remove_special_characters)

### Text stemming

*Stemming is a technique for eliminating affixes from words in order to retrieve the base form.*

In [176]:
#Stemming the text
def simple_stemmer(text):
    ps=nltk.porter.PorterStemmer()
    text= ' '.join([ps.stem(word) for word in text.split()])
    return text
#Apply function on review column
imdb_data['Review']=imdb_data['Review'].apply(simple_stemmer)

### Removing stopwords and normalization

*Stop words are common words with minimal meaning in text analysis and are filtered out during processing. They include words like "a," "an," and "the," but different NLP algorithms may use different lists of stop words.*

In [177]:
#set stopwords to english
stop=set(stopwords.words('english'))
#print(stop)

#removing the stopwords
def remove_stopwords(text, is_lower_case=False):
    tokens = tokenizer.tokenize(text)
    tokens = [token.strip() for token in tokens]
    if is_lower_case:
        filtered_tokens = [token for token in tokens if token not in stopword_list]
    else:
        filtered_tokens = [token for token in tokens if token.lower() not in stopword_list]
    filtered_text = ' '.join(filtered_tokens)
    return filtered_text
#Apply function on review column
imdb_data['Review']=imdb_data['Review'].apply(remove_stopwords)

*Text normalization involves converting text into a consistent canonical form to ensure uniformity before processing or storing, allowing for clear separation of concerns.*

## Bag Of Words Representation

In [178]:
# Split the data into training and test sets
train_reviews, test_reviews, y_train, y_test = train_test_split(imdb_data['Review'], imdb_data['y'], test_size=0.2, random_state=42)

# Assuming you have a function to normalize your text
def normalize_text(text):
    # Example normalization steps
    text = text.lower()  # Convert to lowercase
    # Add more normalization steps here
    return text

# Normalize the text
norm_train_reviews = train_reviews.apply(normalize_text)
norm_test_reviews = test_reviews.apply(normalize_text)


In [179]:
#Count vectorizer for bag of words
cv=CountVectorizer(min_df=0.0,max_df=1,binary=False,ngram_range=(1,3))
#transformed train reviews
cv_train_reviews=cv.fit_transform(norm_train_reviews)
#transformed test reviews
cv_test_reviews=cv.transform(norm_test_reviews)

print('BOW_cv_train:',cv_train_reviews.shape)
print('BOW_cv_test:',cv_test_reviews.shape)
#vocab=cv.get_feature_names()-toget feature names

BOW_cv_train: (815, 83244)
BOW_cv_test: (204, 83244)


## Term Frequency-Inverse Document Frequency representation (TFIDF)

*The term frequency-inverse document frequency (TF-IDF) statistic converts text documents into a matrix of numerical features, reflecting the importance of each word in a document relative to a collection.*

In [180]:
#Tfidf vectorizer
tv=TfidfVectorizer(min_df=0.0,max_df=1,use_idf=True,ngram_range=(1,3))
#transformed train reviews
tv_train_reviews=tv.fit_transform(norm_train_reviews)
#transformed test reviews
tv_test_reviews=tv.transform(norm_test_reviews)
print('Tfidf_train:',tv_train_reviews.shape)
print('Tfidf_test:',tv_test_reviews.shape)

Tfidf_train: (815, 83244)
Tfidf_test: (204, 83244)


### Modelling the dataset

*We will build a Multinomial Naive Bayes model for both Bag of Words and TF-IDF features. This algorithm uses feature vectors where each term's value reflects its frequency or occurrence in the data.*

### Model Training

*Training the model for both the approach both bag of words and tfidf features.*

In [181]:
#training the model
mnb=MultinomialNB()
#fitting the svm for bag of words
mnb_bow=mnb.fit(cv_train_reviews,y_train)
print(mnb_bow)
#fitting the svm for tfidf features
mnb_tfidf=mnb.fit(tv_train_reviews,y_train)
print(mnb_tfidf)


MultinomialNB()
MultinomialNB()


### Model performance on test data

*Model performance by predicting the model for both the approach both bag of words and tfidf features.*

In [182]:
#Predicting the model for bag of words
mnb_bow_predict=mnb.predict(cv_test_reviews)
print(mnb_bow_predict)
#Predicting the model for tfidf features
mnb_tfidf_predict=mnb.predict(tv_test_reviews)
print(mnb_tfidf_predict)

['positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'positive'
 'positive' 'positive' 'positive' 'positive' 'positive' 'posit

### Accuracy of the model

*Accuracy for both the approach both bag of words and tfidf features*

In [183]:
#Accuracy score for bag of words
mnb_bow_score=accuracy_score(y_test,mnb_bow_predict)
print("mnb_bow_score :",mnb_bow_score)
#Accuracy score for tfidf features
mnb_tfidf_score=accuracy_score(y_test,mnb_tfidf_predict)
print("mnb_tfidf_score :",mnb_tfidf_score)

mnb_bow_score : 0.6715686274509803
mnb_tfidf_score : 0.6764705882352942


### Visualizing the classification report

*Visualizing the classification report for both BoW and TFIDF.*



In [184]:
#Classification report for bag of words
mnb_bow_report=classification_report(y_test,mnb_bow_predict,target_names=['positive','negative'])
print(mnb_bow_report)
#Classification report for tfidf features
mnb_tfidf_report=classification_report(y_test,mnb_tfidf_predict,target_names=['positive','negative'])
print(mnb_tfidf_report)

              precision    recall  f1-score   support

    positive       0.00      0.00      0.00        66
    negative       0.67      0.99      0.80       138

    accuracy                           0.67       204
   macro avg       0.34      0.50      0.40       204
weighted avg       0.46      0.67      0.54       204

              precision    recall  f1-score   support

    positive       0.00      0.00      0.00        66
    negative       0.68      1.00      0.81       138

    accuracy                           0.68       204
   macro avg       0.34      0.50      0.40       204
weighted avg       0.46      0.68      0.55       204



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Results Interpretation:

Positive Class:

+ Precision: 0.00
+ Recall: 0.00
+ F1-Score: 0.00
+ Support: 66

The precision, recall, and F1-score for the positive class are all zero, indicating that the model fails to correctly identify any positive reviews. This suggests that the model is significantly biased toward predicting negative sentiments.

Negative Class:

+ Precision: 0.67-0.68
+ Recall: 0.99-1.00
+ F1-Score: 0.80-0.81
+ Support: 138

For the negative class, the precision and recall are high, leading to a strong F1-score. The model is very effective at identifying negative reviews, as it correctly predicts nearly all negative reviews with few false positives.

Overall Metrics:
+ Accuracy: 0.67-0.68
Macro Average:
+ Precision: 0.34
Recall: 0.50
+ F1-Score: 0.40
Weighted Average:
+ Precision: 0.46
Recall: 0.67-0.68
+ F1-Score: 0.54-0.55

The overall accuracy of around 67-68% is heavily influenced by the model's performance on the negative class, given the substantial class imbalance. The macro average values are low, reflecting the poor performance on the positive class, while the weighted average is slightly better due to the higher number of negative samples.

Next Step:
**I will address Class Imbalance:**
I'll implement techniques like oversampling the positive class, undersampling the negative class, or using synthetic data generation to balance the dataset.

In [186]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(solver='liblinear')
lr.fit(tv_train_reviews,y_train) # fit the model
preds = lr.predict(tv_test_reviews) # make predictions

In [187]:
from sklearn.metrics import accuracy_score
accuracy_score(preds,y_test)

0.6764705882352942

I also did try to use logit to build my model and find the accuracy. Its the same across all the models.

### Discussion, Business Insights, and Conclusion

- I think interpreting cultural evolution with movie reviews could be done while performing EDA. I couldn't figure out how I could achieve this by sentiment analysis.
- My accuracy kept going higher as I kept decreasing the rating to generate postive and negative sentiments. (imdb_data['y'] = imdb_data['Rating'].apply(lambda x: 'positive' if x > 5 else 'negative'))
- My model has 0.67 accuracy rate but it doesn't matter when i have a class imbalance. I need to address that for future work.
- I would like to create an interface where i can type a review and choose my model to predict the sentiment.
- I think reviewers now are more aware and appreciative of movies that depict sociatal mistreatment realistically.
- A re-release of this movie now would gain high revenue because of this age and era's love for the vintage classics but it is important to know that some scenes of the movie may be difficult for some viewers to watch based on historical and current treatment of Black people in the United States.